pkg/healthchecker/healthchecker.go

```diff
@@ -0,0 +1,169 @@
+	stsnamespace := s.cluster.Namespace()
+	podname := fmt.Sprintf("%s-%v", stsname, nodeID)
+
+	//we sleep 1 minunte to ensure the pre-stop hook finished
+	time.Sleep(1 * time.Minute)
+	//SKIP KILL because we just restarted????
+	// err := sendKillSignal(ctx, l, logSuffix, podname, nodeID)
+	// if err != nil {
+	// 	return err
+	// }
+
+	//isready check for all pods from sts
+	err := kube.WaitUntilAllStsPodsAreReady(ctx, s.clientset, l, stsname, stsnamespace, 30*time.Minute, 10*time.Minute)
+	if err != nil {
+		return err
+	}
+
+	sts, err := s.clientset.AppsV1().StatefulSets(stsnamespace).Get(ctx, stsname, metav1.GetOptions{})
+	if err != nil {
+		return kube.HandleStsError(err, l, stsname, stsnamespace)
+	}
+	//TODO: add goroutine for each partition
+	for partition := *sts.Spec.Replicas - 1; partition >= 0; partition-- {
+		//we check the metric for all the pods
+		err = s.waitUntilUnderReplicatedMetricIsZero(ctx, l, logSuffix, podname, partition)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+//sendKillSignal it is not necessary in our scenario... we just restarted the node and run drain as pre-stop hook
+func sendKillSignal(ctx context.Context, l logr.Logger, logSuffix, podname string, nodeID int) error {
+	l.V(int(zapcore.DebugLevel)).Info("sendKillSignal", "label", logSuffix, "pod", podname, "nodeID", nodeID)
+	//assuming that crdb process always start with id 1
+	process, err := os.FindProcess(1)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get process cockroach")
+	}
+	//TODO:Check if process.Kill or process.Signal usage
+	err = process.Signal(syscall.Signal(syscall.SIGKILL))
+	if err != nil {
+		return errors.Wrapf(err, "failed to get kill process cockroach")
+	}
+	return nil
+}
+
+func (s *HealthCheckerImpl) waitUntilUnderReplicatedMetricIsZero(ctx context.Context, l logr.Logger, logSuffix, podname string, partition int32) error {
+	f := func() error {
+		return s.checkUnderReplicatedMetric(ctx, l, logSuffix, podname, partition)
+	}
+	b := backoff.NewExponentialBackOff()
+	b.MaxElapsedTime = 3 * time.Minute
+	b.MaxInterval = 5 * time.Second
+	if err := backoff.Retry(f, b); err != nil {
+		return errors.Wrapf(err, "replicas check probe failed for cluster %s", logSuffix)
+	}
+	return nil
+}
+//checkUnderReplicatedMetric uses exec on the pod for now
+func (s *HealthCheckerImpl) checkUnderReplicatedMetric(ctx context.Context, l logr.Logger, logSuffix, podname string, partition int32) error {
+	l.V(int(zapcore.DebugLevel)).Info("checkUnderReplicatedMetric", "label", logSuffix, "podname", podname, "partition", partition)
+	port := strconv.FormatInt(int64(*s.cluster.Spec().HTTPPort), 10)
+	underrepmetric := fmt.Sprintf(underreplicatedmetric, partition)
+	cmd := []string{
+		fmt.Sprintf(cmdunderreplicted, port, underrepmetric),
+	}
+	l.V(int(zapcore.DebugLevel)).Info("get ranges_underreplicated metric", "node", podname)
+	output, stderr, err := kube.ExecInPod(s.scheme, s.config, s.cluster.Namespace(),
```

Comment:
We should be hitting the prometheus endpoint directly.

Reply:
totally agree :)